module/mempool/consensus/execution_tree.go

```diff
@@ -0,0 +1,216 @@
+package consensus
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/forest"
+	"github.com/onflow/flow-go/module/mempool"
+)
+
+// ExecutionTree is a mempool holding receipts, which is aware of the tree structure
+// formed by the results. The mempool supports pruning by height: only results
+// descending from the latest sealed and finalized result are relevant. Hence, we
+// can prune all results for blocks _below_ the latest block with a finalized seal.
+// Results of sufficient height for forks that conflict with the finalized fork are
+// retained. However, such orphaned forks do not grow anymore and their results
+// will be progressively flushed out with increasing sealed-finalized height.
+//
+// Safe for concurrent access. Internally, the mempool utilizes the LevelledForrest.
+// For an in-depth discussion of the core algorithm, see ./Fork-Aware_Mempools.md
+type ExecutionTree struct {
+	sync.RWMutex
+	forest forest.LevelledForest
+	size   uint
+}
+
+// NewReceiptsForest instantiates a ExecutionTree
+func NewReceiptsForest() *ExecutionTree {
+	return &ExecutionTree{
+		RWMutex: sync.RWMutex{},
+		forest:  *forest.NewLevelledForest(),
+		size:    0,
+	}
+}
+
+// AddResult adds an Execution Result to the Execution Tree (without any receipts), in
+// case the result is not already stored in the tree.
+// This is useful for crash recovery:
+// After recovering from a crash, the mempools are wiped and the sealed results will not
+// be stored in the Execution Tree anymore. Adding the result to the tree allows to create
+// a vertex in the tree without attaching any Execution Receipts to it.
+func (et *ExecutionTree) AddResult(result *flow.ExecutionResult, block *flow.Header) error {
+	et.Lock()
+	defer et.Unlock()
+
+	// drop receipts for block heights lower than the lowest height.
+	if block.Height < et.forest.LowestLevel {
+		return nil
+	}
+
+	// sanity check: initial result should be for block
+	if block.ID() != result.BlockID {
+		return fmt.Errorf("receipt is for different block")
+	}
+
+	_, err := et.getEquivalenceClass(result, block)
+	if err != nil {
+		return fmt.Errorf("failed to get equivalence class for result (%x): %w", result.ID(), err)
+	}
+	return nil
+}
+
+// getEquivalenceClass retrieves the Equivalence class for the given result
+// or creates a new one and stores it into the levelled forest
+func (et *ExecutionTree) getEquivalenceClass(result *flow.ExecutionResult, block *flow.Header) (*ReceiptEquivalenceClass, error) {
+	vertex, found := et.forest.GetVertex(result.ID())
+	var receiptsForResult *ReceiptEquivalenceClass
+	if !found {
+		var err error
+		receiptsForResult, err = NewReceiptEquivalenceClass(result, block)
+		if err != nil {
+			return nil, fmt.Errorf("constructing equivalence class for receipt failed: %w", err)
+		}
+		err = et.forest.VerifyVertex(receiptsForResult)
+		if err != nil {
```

Comment:
optional suggestion to make the error message more specific. Vertex and leveled forrest is the implementation detail.